core/slim_sim.cpp

```diff
@@ -6398,115 +6446,142 @@ void SLiMSim::ReadProvenanceTable(tsk_table_collection_t *p_tables, slim_generat
-	std::string slim_record_str(slim_record, slim_record_len);
-	auto j = nlohmann::json::parse(slim_record_str);
-	
-	//std::cout << "Read provenance:\n" << slim_record_str << std::endl;
-	
-	auto file_version_01 = j["file_version"];
-	auto file_version_02 = j["slim"]["file_version"];
-	
+	// New provenance reading code, using the JSON for Modern C++ library (json.hpp); 
+    
 	std::string model_type_str;
 	long long gen_ll;
-	
-	if (file_version_01.is_string() && (file_version_01 == "0.1"))
-	{
-		// We actually don't have any chance of being able to read SLiM 3.0 .trees files in, I guess;
-		// all the new individuals table stuff, the addition of the population table, etc., mean that
-		// it would just be a huge headache to try to do this.  So let's throw an error up front.
-		EIDOS_TERMINATION << "ERROR (SLiMSim::ReadProvenanceTable): file_version is 0.1 in .trees file; this file cannot be read.  SLiM 3.1 and later cannot read saved .trees files from prior versions of SLiM; sorry." << EidosTerminate();
-		
-		/*
-		try {
-			model_type_str = j["model_type"];
-			gen_ll = j["generation"];
-			//rem_count_ll = j["remembered_node_count"];	// no longer using this key
-		}
-		catch (...)
-		{
-			EIDOS_TERMINATION << "ERROR (SLiMSim::ReadProvenanceTable): error reading provenance value (file_version 0.1); this file cannot be read." << EidosTerminate();
-		}
-		 */
-	}
-	else if (file_version_02.is_string())
-	{
-		// File version 0.2 was before the addition of nucleotides, so MutationMetadataRec_PRENUC will need to be used
-		// File version 0.3 is the current format, with nucleotides, using MutationMetadataRec
-		if (file_version_02 == SLIM_TREES_FILE_VERSION_PRENUC)
-			*p_file_version = 2;
-		else if (file_version_02 == SLIM_TREES_FILE_VERSION_POSTNUC)
-			*p_file_version = 3;
-		else if (file_version_02 == SLIM_TREES_FILE_VERSION)
-			*p_file_version = 4;
-		else
-			EIDOS_TERMINATION << "ERROR (SLiMSim::ReadProvenanceTable): this .trees file was generated by an unrecognized version of SLiM or pyslim; this file cannot be read." << EidosTerminate();
-		
-		try {
-			model_type_str = j["parameters"]["model_type"];
-			gen_ll = j["slim"]["generation"];
-			//rem_count_ll = j["slim"]["remembered_node_count"];	// no longer using this key
-		}
-		catch (...)
-		{
-			EIDOS_TERMINATION << "ERROR (SLiMSim::ReadProvenanceTable): error reading provenance value (file_version " << file_version_02 << "); this file cannot be read." << EidosTerminate();
-		}
-	}
-	else
-	{
-		EIDOS_TERMINATION << "ERROR (SLiMSim::ReadProvenanceTable): missing or corrupted file version; this file cannot be read." << EidosTerminate();
-	}
+
+    ////////////
+    // Format 0.5: using top-level metadata
+    try {
+        // Note: we *could* parse the metadata schema (which is json),
+        // but instead we'll just try parsing the metadata.
```

Comment:
not sure I grasp the distinction.  what would it mean to parse the metadata schema?

Reply:
Well, the distinction makes more sense for the tables - e.g., the mutation metadata. The metadata schema says "here's what's in the metadata". So, it's a way of letting the tree sequence carry along with it how to interpret the stuff we stick in to the metadata itself. This will tell the python code how to decode the metadata and also will let python validate new metadata - so when you try to add a new row to the mutation table in python, it will check if the metadata you've specified fits within the metadata schema. This is super useful... but, maybe not for the top-level metadata, of which there is only one of, and carries around names with all the values anyhow. I guess it's useful to write down the schema, since it's one way of saying precisely what is required in a valid SLiM tree sequence metadata entry, but I don't think the tree sequence itself even needs to carry it around.

In fact, someone else (eg pyslim) could add new keys to the tree sequence metadata and that wouldn't get in our way. But then what would they do with the schema? I think maybe we don't even want to put it in? I'll see what Ben Jeffries says.

Reply:
Hmm.  So for metadata on things like mutations, the schema is a binding contract: exactly these pieces of data must be present, in exactly this order, in this data type.  And that's because that metadata is raw bytes, not JSON.  But for the tree sequence top-level metadata, that is stored as JSON, and so the schema has a different meaning, semantically; it's a *minimum*, stating which keys need to be present, but other keys could also be added, and that's harmless because it's JSON.  Is that correct?  Is the top-level TS metadata JSON by official policy, or does SLiM just happen to store its "raw bytes" in JSON format, but tskit makes no guarantee that that metadata will be JSON in general?  If it's JSON by official tskit policy, then it seems like it would be good and useful for a schema to be able to be attached to it, as long as it's recognized that that schema is a minimum, in the sense I just said.  In fact, maybe it would be useful to be able to attach more than one schema; SLiM could have its standard schema there, and if pyslim added additional keys to the JSON, it could also attach a pyslim schema that would specify/describe those keys, without modifying the SLiM schema.  Is that crazy?  Anyway, yes, I'd like to hear what Ben Jeffrey says.

Reply:
Right: currently, table metadata can be either json or raw bytes; and if it's raw bytes then everything's got to be "just like this", as you say. If it's json it can be more flexible. (Since a schema can specify "only these things please and in this order" or can say "at least this stuff" etcetera.) The top-level metadata is supposed to be json, and I think that's written down somewhere, yes. But - what extra information is conveyed in the schema that's not in the metadata itself?

Reply:
> But - what extra information is conveyed in the schema that's not in the metadata itself?

Hmm.  Well, not just "here are the keys that are present", but "here are the keys that are *expected* to be present", "here is the standard that this document claims to conform to", etc., right?  From which one can deduce things like "This document does not even claim to be SLiM-conformant" versus "This document claims to be SLiM-conformant but is not", "This document is conformant to an earlier version of SLiM but claims to be conformant to a later version", "This document has additional keys present beyond those required to be conformant", etc.  I don't know that we have any immediate plans to *do* anything with that information, but it doesn't seem useless in principle...?  But let's see what the other Ben (the imposter Ben!) says about this, I guess.  I'm not sure that I feel strongly either way, but I'm also not sure I appreciate all the implications.

Reply:
Well, what matters is whether it *is* conformant. Adding a claim that it's conformant to something seems only helpful if it helps us decide how to parse it. These other things you mention (eg 'does not claim to be conformant' or 'claims to be X but is actually Y') are not information about the tree sequence, but rather information about the *claim* itself. (and so, not new information about the tree sequence)

Reply:
Yes; it's not clear to me that that means it's useless information, though.  For example: I can imagine deciding that SLiM should only open TS files that not only *are* conformant to SLiM's requirements, but that also explicitly *claim* to be conformant.  The claim is an expression of intent; it says "the producer of this file intends for it to be SLiM-conformant".  And maybe that is a meaningful and important claim.  Some other producer of TS files might not *intend* for its files to be SLiM-conformant, but they might end up having the TS-level metadata just copied over in bulk, as a part of a processing step, say.  Such a producer might wish to copy that TS-level metadata, while removing the *claim* that it is conformant (because it makes no attempt to be conformant, and perhaps doesn't even have any awareness at all of what SLiM is or what SLiM-conformance would entail).  Copying the metadata while omitting the claim would be a way to express that sort of intent.  Again, I don't know if that's a good idea or not, in practice; but I'm pushing back against what you seem to be saying, that there can be no value at all to such claims as far as the actual parsing of the tree sequence is concerned.

Reply:
I agree with the sentiment, but I don't think that the metadata schema is the right way to signal that. Anyhow, we'll see what other-Ben says.

Reply:
Ben points out that you can put a `description` field into the metadata schema, which gives the user substantially more information about what's there. So, that's good.